// @freesewing/models/dist/index.mjs
/**
 * @freesewing/models | v2.17.0
 * Body measurements data for a range of default sizes
 * (c) 2021 Joost De Cock <joost@joost.at> (https://github.com/joostdecock)
 * @license MIT
 */
import neckstimate from '@freesewing/utils/neckstimate';

var measurements = {
  menswear: [
    'ankle',
    'biceps',
    'chest',
    'crossSeam',
    'crossSeamFront',
    'crotchDepth',
    'head',
    'heel',
    'hips',
    'hpsToBust',
    'hpsToWaistBack',
    'inseam',
    'knee',
    'neck',
    'seat',
    'seatBack',
    'shoulderSlope',
    'shoulderToElbow',
    'shoulderToShoulder',
    'shoulderToWrist',
    'upperLeg',
    'waist',
    'waistBack',
    'waistToFloor',
    'waistToHips',
    'waistToKnee',
    'waistToSeat',
    'waistToUpperLeg',
    'wrist',
  ],
  womenswear: [
    'ankle',
    'biceps',
    'bustFront',
    'bustSpan',
    'chest',
    'crossSeam',
    'crossSeamFront',
    'crotchDepth',
    'head',
    'heel',
    'highBust',
    'highBustFront',
    'hips',
    'hpsToBust',
    'hpsToWaistBack',
    'hpsToWaistFront',
    'inseam',
    'knee',
    'neck',
    'seat',
    'seatBack',
    'shoulderSlope',
    'shoulderToElbow',
    'shoulderToShoulder',
    'shoulderToWrist',
    'underbust',
    'upperLeg',
    'waist',
    'waistBack',
    'waistToFloor',
    'waistToHips',
    'waistToKnee',
    'waistToSeat',
    'waistToUnderbust',
    'waistToUpperLeg',
    'wrist',
  ],
};

var sizes = {
  menswear: [32, 34, 36, 38, 40, 42, 44, 46, 48, 50],
  womenswear: [28, 30, 32, 34, 36, 38, 40, 42, 44, 46],
};

const withBreasts = {};
const withoutBreasts = {};

for (let s of sizes.womenswear) {
  withBreasts['size' + s] = {};
  for (let m of measurements.womenswear) {
    withBreasts['size' + s][m] = neckstimate(s * 10, m, true);
  }
}

for (let s of sizes.menswear) {
  withoutBreasts['size' + s] = {};
  for (let m of measurements.menswear) {
    withoutBreasts['size' + s][m] = neckstimate(s * 10, m, false);
  }
}

export { measurements, sizes, withBreasts, withoutBreasts };
