// @freesewing/components/SampleConfigurator/index.js
/**
 * @freesewing/components | v2.17.0
 * A collection of React components for FreeSewing web UIs
 * (c) 2021 Joost De Cock <joost@joost.at> (https://github.com/joostdecock)
 * @license MIT
 */
'use strict';

var React = require('react');
var reactIntl = require('react-intl');
var models = require('@freesewing/models');
var neckstimate = require('@freesewing/utils/neckstimate');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var neckstimate__default = /*#__PURE__*/_interopDefaultLegacy(neckstimate);

function _typeof(obj) {
  "@babel/helpers - typeof";

  if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {
    _typeof = function (obj) {
      return typeof obj;
    };
  } else {
    _typeof = function (obj) {
      return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj;
    };
  }

  return _typeof(obj);
}

function _slicedToArray(arr, i) {
  return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest();
}

function _arrayWithHoles(arr) {
  if (Array.isArray(arr)) return arr;
}

function _iterableToArrayLimit(arr, i) {
  var _i = arr && (typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]);

  if (_i == null) return;
  var _arr = [];
  var _n = true;
  var _d = false;

  var _s, _e;

  try {
    for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) {
      _arr.push(_s.value);

      if (i && _arr.length === i) break;
    }
  } catch (err) {
    _d = true;
    _e = err;
  } finally {
    try {
      if (!_n && _i["return"] != null) _i["return"]();
    } finally {
      if (_d) throw _e;
    }
  }

  return _arr;
}

function _unsupportedIterableToArray(o, minLen) {
  if (!o) return;
  if (typeof o === "string") return _arrayLikeToArray(o, minLen);
  var n = Object.prototype.toString.call(o).slice(8, -1);
  if (n === "Object" && o.constructor) n = o.constructor.name;
  if (n === "Map" || n === "Set") return Array.from(o);
  if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen);
}

function _arrayLikeToArray(arr, len) {
  if (len == null || len > arr.length) len = arr.length;

  for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i];

  return arr2;
}

function _nonIterableRest() {
  throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
}

function _createForOfIteratorHelper(o, allowArrayLike) {
  var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"];

  if (!it) {
    if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") {
      if (it) o = it;
      var i = 0;

      var F = function () {};

      return {
        s: F,
        n: function () {
          if (i >= o.length) return {
            done: true
          };
          return {
            done: false,
            value: o[i++]
          };
        },
        e: function (e) {
          throw e;
        },
        f: F
      };
    }

    throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
  }

  var normalCompletion = true,
      didErr = false,
      err;
  return {
    s: function () {
      it = it.call(o);
    },
    n: function () {
      var step = it.next();
      normalCompletion = step.done;
      return step;
    },
    e: function (e) {
      didErr = true;
      err = e;
    },
    f: function () {
      try {
        if (!normalCompletion && it.return != null) it.return();
      } finally {
        if (didErr) throw err;
      }
    }
  };
}

var OptionGroup = function OptionGroup(props) {
  return /*#__PURE__*/React__default['default'].createElement(React__default['default'].Fragment, null, props.options.map(function (name) {
    var output = [];

    if (_typeof(name) === 'object') {
      // Subgroup
      for (var _i = 0, _Object$keys = Object.keys(name); _i < _Object$keys.length; _i++) {
        var subGroup = _Object$keys[_i];
        output.push( /*#__PURE__*/React__default['default'].createElement("h5", {
          key: subGroup + '-title'
        }, /*#__PURE__*/React__default['default'].createElement(reactIntl.FormattedMessage, {
          id: 'optiongroups.' + subGroup
        })));
        var children = [];

        var _iterator = _createForOfIteratorHelper(name[subGroup]),
            _step;

        try {
          var _loop = function _loop() {
            var option = _step.value;
            children.push( /*#__PURE__*/React__default['default'].createElement("li", null, /*#__PURE__*/React__default['default'].createElement("a", {
              href: "#logo",
              onClick: function onClick() {
                return props.sampleOption(option);
              },
              className: props.activeOption === option ? 'active' : ''
            }, /*#__PURE__*/React__default['default'].createElement(reactIntl.FormattedMessage, {
              id: 'options.' + props.config.name + '.' + option + '.title'
            }))));
          };

          for (_iterator.s(); !(_step = _iterator.n()).done;) {
            _loop();
          }
        } catch (err) {
          _iterator.e(err);
        } finally {
          _iterator.f();
        }

        output.push( /*#__PURE__*/React__default['default'].createElement("ul", {
          style: {
            paddingLeft: '1rem'
          }
        }, children));
      }
    } else output.push( /*#__PURE__*/React__default['default'].createElement("li", null, /*#__PURE__*/React__default['default'].createElement("a", {
      href: "#logo",
      onClick: function onClick() {
        return props.sampleOption(name);
      },
      className: props.activeOption === name ? 'active' : ''
    }, /*#__PURE__*/React__default['default'].createElement(reactIntl.FormattedMessage, {
      id: 'options.' + props.config.name + '.' + name + '.title'
    }))));

    return output;
  }));
};

var OptionGroup$1 = reactIntl.injectIntl(OptionGroup);

var PatternOptions = function PatternOptions(props) {
  var renderGroup = function renderGroup(group) {
    var output = [];
    var children = /*#__PURE__*/React__default['default'].createElement("ul", null, /*#__PURE__*/React__default['default'].createElement(OptionGroup$1, {
      key: group + '-group',
      units: props.units,
      config: props.config,
      options: props.config.optionGroups[group],
      sampleOption: props.sampleOption,
      activeOption: props.activeOption
    }));
    output.push( /*#__PURE__*/React__default['default'].createElement("li", {
      key: group + '-ghead'
    }, /*#__PURE__*/React__default['default'].createElement("h6", null, /*#__PURE__*/React__default['default'].createElement(reactIntl.FormattedMessage, {
      id: 'optiongroups.' + group
    })), children));
    return output;
  };

  return /*#__PURE__*/React__default['default'].createElement("ul", null, Object.keys(props.config.optionGroups).map(function (group) {
    return renderGroup(group);
  }));
};

var SampleConfigurator = function SampleConfigurator(props) {
  var _useState = React.useState(),
      _useState2 = _slicedToArray(_useState, 2),
      type = _useState2[0],
      setType = _useState2[1];

  var _useState3 = React.useState(),
      _useState4 = _slicedToArray(_useState3, 2),
      instance = _useState4[0],
      setInstance = _useState4[1];

  var sampleOption = function sampleOption(option) {
    setType('option');
    setInstance(option);
    props.updateGist({
      type: 'option',
      option: option
    }, 'settings', 'sample');
  };

  var sampleMeasurement = function sampleMeasurement(measurement) {
    setType('measurement');
    setInstance(measurement);
    props.updateGist({
      type: 'measurement',
      measurement: measurement
    }, 'settings', 'sample');
  };

  var sampleModels = function sampleModels(models) {
    setType('models');
    props.updateGist({
      type: 'models',
      models: models
    }, 'settings', 'sample');
  };

  var antMan = {
    ant: {},
    b: {},
    c: {},
    man: {}
  };
  var antWoman = {
    ant: {},
    b: {},
    c: {},
    woman: {}
  };

  for (var m in models.withoutBreasts.size42) {
    var val = neckstimate__default['default'](420, m, false);
    antMan.ant[m] = val / 10;
    antMan.b[m] = val / 5;
    antMan.c[m] = val / 2;
    antMan.man[m] = val;
  }

  for (var _m in models.withBreasts.size36) {
    var _val = neckstimate__default['default'](360, _m, true);

    antWoman.ant[_m] = _val / 10;
    antWoman.b[_m] = _val / 5;
    antWoman.c[_m] = _val / 2;
    antWoman.woman[_m] = _val;
  }

  return /*#__PURE__*/React__default['default'].createElement("ul", null, /*#__PURE__*/React__default['default'].createElement("li", null, /*#__PURE__*/React__default['default'].createElement("h5", null, /*#__PURE__*/React__default['default'].createElement(reactIntl.FormattedMessage, {
    id: "app.patternOptions"
  })), /*#__PURE__*/React__default['default'].createElement(PatternOptions, {
    config: props.config,
    gist: props.gist,
    sampleOption: sampleOption,
    activeOption: type === 'option' ? instance : null
  })), /*#__PURE__*/React__default['default'].createElement("li", null, /*#__PURE__*/React__default['default'].createElement("h5", null, /*#__PURE__*/React__default['default'].createElement(reactIntl.FormattedMessage, {
    id: "app.measurements"
  })), /*#__PURE__*/React__default['default'].createElement("ul", null, props.config.measurements.map(function (m) {
    return /*#__PURE__*/React__default['default'].createElement("li", {
      key: m
    }, /*#__PURE__*/React__default['default'].createElement("a", {
      href: "#logo",
      onClick: function onClick() {
        return sampleMeasurement(m);
      },
      className: type === 'measurement' && instance === m ? 'active' : ''
    }, /*#__PURE__*/React__default['default'].createElement(reactIntl.FormattedMessage, {
      id: 'measurements.' + m
    })));
  }))), /*#__PURE__*/React__default['default'].createElement("li", null, /*#__PURE__*/React__default['default'].createElement("h5", null, /*#__PURE__*/React__default['default'].createElement(reactIntl.FormattedMessage, {
    id: "app.people"
  })), /*#__PURE__*/React__default['default'].createElement("ul", null, /*#__PURE__*/React__default['default'].createElement("li", null, /*#__PURE__*/React__default['default'].createElement("a", {
    href: "#logo",
    onClick: function onClick() {
      sampleModels(models.withBreasts);
      setType('models');
      setInstance('withBreasts');
    },
    className: type === 'models' && instance === 'withBreasts' ? 'active' : ''
  }, /*#__PURE__*/React__default['default'].createElement(reactIntl.FormattedMessage, {
    id: "app.withBreasts"
  }))), /*#__PURE__*/React__default['default'].createElement("li", null, /*#__PURE__*/React__default['default'].createElement("a", {
    href: "#logo",
    onClick: function onClick() {
      sampleModels(models.withoutBreasts);
      setType('models');
      setInstance('withoutBreasts');
    },
    className: type === 'models' && instance === 'withoutBreasts' ? 'active' : ''
  }, /*#__PURE__*/React__default['default'].createElement(reactIntl.FormattedMessage, {
    id: "app.withoutBreasts"
  }))), /*#__PURE__*/React__default['default'].createElement("li", null, /*#__PURE__*/React__default['default'].createElement("a", {
    href: "#logo",
    onClick: function onClick() {
      sampleModels(antWoman);
      setType('models');
      setInstance('antWoman');
    },
    className: type === 'models' && instance === 'antWoman' ? 'active' : ''
  }, "Antperson (with breasts)")), /*#__PURE__*/React__default['default'].createElement("li", null, /*#__PURE__*/React__default['default'].createElement("a", {
    href: "#logo",
    onClick: function onClick() {
      sampleModels(antMan);
      setType('models');
      setInstance('antMan');
    },
    className: type === 'models' && instance === 'antMan' ? 'active' : ''
  }, "Antperson (without breasts)")))));
};

module.exports = SampleConfigurator;
